### PR TITLE
Fix clippy warnings

### LINKED ISSUE #126: Fix clippy warnings for all optional features
Warnings can be found with commands 

- `cargo clippy --features compact_filters --no-default-features`

Also need to fix:

- wallet/utils.rs#106, fix and remove `#[allow(dead_code)]`
- database/any.rs#325, fix and remove `#[allow(irrefutable_let_patterns)]`

After warnings are fixed enable clippy checks by default in `cont_integration.yml` for the `build-test` job:
   
- for `strategy.matrix.include` remove any with: 
   `clippy: false`
   
- and in the `clippy` step remove:
    `if: ${{ matrix.clippy == true }}`

### PR DESCRIPTION
**Pull request by tcharding**
### Description

As described in issue #126 we have a bunch of clippy warnings. This PR clears the warnings. 

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [ ] I've added tests for the new feature
* [ ] I've added docs for the new feature
* [ ] I've updated `CHANGELOG.md`

#### Bugfixes:

* [ ] This pull request breaks the existing API
* [ ] I've added tests to reproduce the issue which are now passing
* [x] I'm linking the issue being fixed by this PR

Fixes: #126 

(I think this resolves the issue, the CI description looks stale but perhaps I'm misunderstanding it, please review the issue description to see if I have missed anything).

Thanks

### CODE REVIEW

**Comment by afilini** [comment]
Concept ACK, you've probably missed something in `testutils-macros/src/lib.rs` where the tests are generated for `Blockchain` backends. That's causing the CI failure in the Electrum test.

**Comment by notmandatory** [comment]
I think most of these fixes are for the `nightly` version of clippy.  As long as the changes `experimental` recommends are compatible with `stable` clippy they're still good to fix, but you just needs to be double for each of the features in the `build-test` cont_integration.yml workflow: 
```
cargo +stable clippy --features <features> -- -D warnings
```

File: src/wallet/signer.rs [hunk 1]
```diff
@@ -617,22 +636,14 @@ mod signers_container_tests {
...
-        );
-        assert!(
-            matches!(signers.find(id2), Some(signer) if Arc::as_ptr(&signer2) == Arc::as_ptr(signer))
-        );
-        assert!(
-            matches!(signers.find(id3.clone()), Some(signer) if Arc::as_ptr(&signer3) == Arc::as_ptr(signer))
-        );
+        assert!(matches!(signers.find(id1), Some(signer) if is_equal(signer, &signer1)));
+        assert!(matches!(signers.find(id2), Some(signer) if is_equal(signer, &signer2)));
+        assert!(matches!(signers.find(id3.clone()), Some(signer) if is_equal(signer, &signer3)));
```
Comment (LLFourn) [inline src/wallet/signer.rs:641]: ❓ Can't we just make `dyn Signer` `Eq`?

@afilini is there some coherent way we can identify a signer i.e. by it's xprv/xpub or something more clever?

**Comment by tcharding** [comment]
> I think most of these fixes are for the `nightly` version of clippy. 

Ok, I think I need to learn how to correctly set up my dev environment for new projects before attempting to contribute :)

> As long as the changes `experimental` recommends are compatible with `stable` clippy they're still good to fix, but you just needs to be double for each of the features in the `build-test` cont_integration.yml workflow:

Closing for now.

> ```
> cargo +stable clippy --features <features> -- -D warnings
> ```

Thanks for schooling me. This was not the only project where I was wondering why so many clippy warnings were there on master - turns out its me :( 

I will pull out the relevant changes and re-open a new PR later on. Sorry for the noise.

**Comment by notmandatory** [comment]
Hey no problem, rust tooling is great but does have it's own learning curve to deal with. For this project we run our CI build/test pipeline against minimum version `1.45.0` and current `stable`. You can set your default rust toolchain version to `stable` with the command:
```
rustup default stable
```

File: src/wallet/signer.rs (hunk 1 above)
Comment (afilini) [reply src/wallet/signer.rs:641]: Yeah, makes sense. I would add an `.id()` method to `Signer` that returns a `SignerId`. Then we can implement `Eq` by checking the id returned by both signers.

Let me open an issue for this

**Pull request closed without merging**